app/controllers/api/base_controller.rb

```diff
@@ -260,6 +247,67 @@ def action_permission
       end
     end
 
+    def parent_permission(child_permission)
+      case child_permission.to_s
+      when 'create', 'destroy'
+        'edit'
+      when 'edit', 'view'
+        'view'
+      else
+        raise ::Foreman::Exception.new(N_("unknown parent permission for %s"), "#{params[:controller]}##{child_permission}")
+      end
+    end
+
+    def parent_resource_details
+      classes_arr = params.map do |key, value|
+        md = key.match(/(\w+)_id$/)
+        if md
+          values = [md[1], resource_class_for(resource_name_for(md[1])), value]
+        else
+          values = [nil, nil, nil]
+        end
+        values
```

Comment:
I hope that this will be refactored by extracting but just for the records, you don't need to assign and later return values, since the result of block would be the values array already (last expression result)